src/iter/update.rs

```diff
@@ -0,0 +1,295 @@
+}
+
+impl<'f, P, F> Producer for UpdateProducer<'f, P, F>
+    where P: Producer,
+          F: Fn(&mut P::Item) + Send + Sync,
+{
+    type Item = P::Item;
+    type IntoIter = UpdateSeq<P::IntoIter, &'f F>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        UpdateSeq {
+            base: self.base.into_iter(),
+            update_op: self.update_op,
+        }
+    }
+
+    fn min_len(&self) -> usize {
+        self.base.min_len()
+    }
+    fn max_len(&self) -> usize {
+        self.base.max_len()
+    }
+
+    fn split_at(self, index: usize) -> (Self, Self) {
+        let (left, right) = self.base.split_at(index);
+        (UpdateProducer {
+             base: left,
+             update_op: self.update_op,
+         },
+         UpdateProducer {
+             base: right,
+             update_op: self.update_op,
+         })
+    }
+}
+
+
+/// ////////////////////////////////////////////////////////////////////////
+/// Consumer implementation
+
+struct UpdateConsumer<'f, C, F: 'f> {
+    base: C,
+    update_op: &'f F,
+}
+
+impl<'f, C, F> UpdateConsumer<'f, C, F> {
+    fn new(base: C, update_op: &'f F) -> Self {
+        UpdateConsumer {
+            base: base,
+            update_op: update_op,
+        }
+    }
+}
+
+impl<'f, T, C, F> Consumer<T> for UpdateConsumer<'f, C, F>
+    where C: Consumer<T>,
+          F: Fn(&mut T) + Send + Sync,
+{
+    type Folder = UpdateFolder<'f, C::Folder, F>;
+    type Reducer = C::Reducer;
+    type Result = C::Result;
+
+    fn split_at(self, index: usize) -> (Self, Self, Self::Reducer) {
+        let (left, right, reducer) = self.base.split_at(index);
+        (UpdateConsumer::new(left, self.update_op), UpdateConsumer::new(right, self.update_op), reducer)
+    }
+
+    fn into_folder(self) -> Self::Folder {
+        UpdateFolder {
+            base: self.base.into_folder(),
+            update_op: self.update_op,
+        }
+    }
+
+    fn full(&self) -> bool {
+        self.base.full()
+    }
+}
+
+impl<'f, T, C, F> UnindexedConsumer<T> for UpdateConsumer<'f, C, F>
+    where C: UnindexedConsumer<T>,
+          F: Fn(&mut T) + Send + Sync,
+{
+    fn split_off_left(&self) -> Self {
+        UpdateConsumer::new(self.base.split_off_left(), &self.update_op)
+    }
+
+    fn to_reducer(&self) -> Self::Reducer {
+        self.base.to_reducer()
+    }
+}
+
+struct UpdateFolder<'f, C, F: 'f> {
+    base: C,
+    update_op: &'f F,
+}
+
+impl<'f, T, C, F> Folder<T> for UpdateFolder<'f, C, F>
+    where C: Folder<T>,
+          F: Fn(& mut T)
+{
+    type Result = C::Result;
+
+    fn consume(self, mut item: T) -> Self {
+        (self.update_op)(&mut item);
+
+        UpdateFolder {
+            base: self.base.consume(item),
+            update_op: self.update_op,
+        }
+    }
+
+    fn complete(self) -> C::Result {
+        self.base.complete()
+    }
+
+    fn full(&self) -> bool {
+        self.base.full()
+    }
+}
+
+/// Standard Update adaptor, based on `itertools::adaptors::Update` 
```

Comment:
Did you consider using itertools directly?  We wouldn't want a public dependency, but this is behind the scenes.  But the implementation is pretty straightforward, so I'm just wondering what you think.

Reply:
I did think about that while reading the issue. I chose to replicate the implementation because `interleave` did the same, although with additional reasons.

If there is willingness to depend on `itertools` in future versions, I suppose this adaptor would take advantage of that. Although it's just my opinion, leaving these adaptors  behind a feature gate to keep `itertools` an optional dependency feels a bit unnecessary.